In [19]:
from population import generate_init
from character import CharacterType
#from src.stats import random_stats

N = 20
maxStatsValue = 150
type = CharacterType.ARCHER

#equipment_data = random_stats(N, maxStatsValue)
poblacion_inicial = generate_init(N,maxStatsValue,type)

#poblacion_inicial


In [2]:
#Añado las columnas de desempeño relativo y acumulado
from selection import add_relative_acumulate
poblacion_rel_acu = add_relative_acumulate(poblacion_inicial)
poblacion_rel_acu

strength  agility  expertise  resistance  life  height  \
17        39       12         64          20    15     1.6   
8         71       16         37          15    11     1.3   
1         36       12         38          23    41     1.6   
19        24       27         51          11    37     1.4   
3         50        6         21          64     9     1.4   
11        74        9         14          49     4     1.7   
13        17       17         67          42     7     1.3   
9         17        8         59          52    14     1.7   
16        14       66         34           3    33     1.8   
2         35       56          9          29    21     1.9   
10        21       15         26          20    68     1.5   
0         15       70         17           9    39     1.8   
14        13       80         17          26    14     1.9   
5          7       75         43          18     7     2.0   
15         4        8         97          16    25     1.8   
18        38       28          4          10    70     1.6   
7         15       28         14           4    89     1.9   
4         78        9          3          23    37     1.6   
12        16       26          4          70    34     1.4   
6          5        7         18          11   109     1.5   

                    type  performance  performance_relative  \
17  CharacterType.ARCHER   690.892226              0.141541   
8   CharacterType.ARCHER   680.930057              0.139500   
1   CharacterType.ARCHER   427.263203              0.087532   
19  CharacterType.ARCHER   398.638298              0.081668   
3   CharacterType.ARCHER   322.803639              0.066132   
11  CharacterType.ARCHER   307.279376              0.062951   
13  CharacterType.ARCHER   305.154920              0.062516   
9   CharacterType.ARCHER   295.225256              0.060482   
16  CharacterType.ARCHER   198.252281              0.040615   
2   CharacterType.ARCHER   197.785852              0.040520   
10  CharacterType.ARCHER   196.043193              0.040163   
0   CharacterType.ARCHER   134.503694              0.027555   
14  CharacterType.ARCHER   122.181167              0.025031   
5   CharacterType.ARCHER   117.570540              0.024086   
15  CharacterType.ARCHER   110.268677              0.022590   
18  CharacterType.ARCHER   100.112252              0.020510   
7   CharacterType.ARCHER    96.981614              0.019868   
4   CharacterType.ARCHER    94.169304              0.019292   
12  CharacterType.ARCHER    44.801461              0.009178   
6   CharacterType.ARCHER    40.364827              0.008269   

    performance_accumulated  
17                 0.141541  
8                  0.281041  
1                  0.368573  
19                 0.450241  
3                  0.516372  
11                 0.579324  
13                 0.641840  
9                  0.702321  
16                 0.742937  
2                  0.783457  
10                 0.823619  
0                  0.851175  
14                 0.876205  
5                  0.900292  
15                 0.922882  
18                 0.943392  
7                  0.963260  
4                  0.982552  
12                 0.991731  
6                  1.000000

#### Seleccion de k indiduos
A*(método1) + (1-A)*(método2)

In [3]:
import numpy as np
#Dado un K
#Si quiero un 40% de la selección con metodo1
K = 10
m1_perc = 0.3
N1 = int(np.round(K * m1_perc))
N2 = K - N1
#N1

In [4]:
from selection import elitist_selection
nueva_poblacion_e = elitist_selection(poblacion_rel_acu,N1)
#nueva_poblacion_e

In [5]:
from selection import roulette_wheel_selection
nueva_poblacion_rul = roulette_wheel_selection(poblacion_rel_acu, N2)
#nueva_poblacion_rul

In [6]:
import pandas as pd
selection_total = pd.concat([nueva_poblacion_e,nueva_poblacion_rul])
selection_total

strength agility expertise resistance life height                  type  \
0       39      12        64         20   15    1.6  CharacterType.ARCHER   
1       71      16        37         15   11    1.3  CharacterType.ARCHER   
2       36      12        38         23   41    1.6  CharacterType.ARCHER   
0       36      12        38         23   41    1.6  CharacterType.ARCHER   
1       78       9         3         23   37    1.6  CharacterType.ARCHER   
2       36      12        38         23   41    1.6  CharacterType.ARCHER   
3       15      70        17          9   39    1.8  CharacterType.ARCHER   
4       17      17        67         42    7    1.3  CharacterType.ARCHER   
5       13      80        17         26   14    1.9  CharacterType.ARCHER   
6       21      15        26         20   68    1.5  CharacterType.ARCHER   

  performance performance_relative performance_accumulated  
0  690.892226             0.141541                0.141541  
1  680.930057               0.1395                0.281041  
2  427.263203             0.087532                0.368573  
0  427.263203             0.087532                0.368573  
1   94.169304             0.019292                0.982552  
2  427.263203             0.087532                0.368573  
3  134.503694             0.027555                0.851175  
4   305.15492             0.062516                 0.64184  
5  122.181167             0.025031                0.876205  
6  196.043193             0.040163                0.823619

In [7]:
from genes import  encode_genes

cromosomas = encode_genes(selection_total)
#cromosomas

In [8]:
from crossover import single_point_crossover
new_individuals = single_point_crossover(cromosomas)
#new_individuals


In [9]:
#from genes import decode_genes
#cromosomas_decoded = decode_genes(new_individuals)
#cromosomas_decoded

In [10]:
#from crossover import valid_offspring
#offspings, offspring_ok = valid_offspring(cromosomas_decoded)

#print(f"Cantidad de hijos validos: {offspings}")
#offspring_ok


- Cantidad de hijos validos: 4
-     strength  agility expertise	resistance	life  height  type	                performance	 performance_relative   performance_accumulated
- 5	  2	        30	    3	        6	        109	  2.0	  CharacterType.ARCHER	6.277699	 0.001625	            1.0
- 6	  50	    18	    3	        18	        61	  1.7	  CharacterType.ARCHER	119.533975	 0.030933	            0.881345
- 9	  4	        39	    92	        4	        11	  1.4	  CharacterType.ARCHER	106.010218	 0.027433	            0.908778
- 10  4	        39	    92	        4	        11	  1.4	  CharacterType.ARCHER	106.010218	 0.027433	            0.908778

In [11]:
from crossover import two_point_crossover

new_individuals2 = two_point_crossover(cromosomas)
#new_individuals2



In [12]:
#from genes import decode_genes
#cromosomas_decoded2 = decode_genes(new_individuals2)
#cromosomas_decoded2

In [13]:
#from crossover import valid_offspring
#offspings, offspring_ok = valid_offspring(cromosomas_decoded2)

#print(f"Cantidad de hijos validos: {offspings}")
#offspring_ok

In [14]:
from crossover import valid_chromosome
cromosomas_ok = valid_chromosome(new_individuals2)
#cromosomas_ok

In [15]:
#from genes import decode_genes
#cromosomas_decoded2 = decode_genes(cromosomas_ok)
#cromosomas_decoded2

In [16]:
from mutation import one_gene_mutation

# Example usage:

mutation_rate = 0.1
result = one_gene_mutation(cromosomas_ok, mutation_rate)
result


strength   agility expertise resistance      life height
2  00100100  00001100  00100110   00010111  00101001    011
3  00100100  00001100  00100110   00010111  00101001    011
4  01001110  00001001  00000011   00010111  00100101    011
5  00100100  00001100  00100110   00010111  00101001    011

In [17]:
from crossover import valid_chromosome
cromosomas_ok = valid_chromosome(result)
cromosomas_ok

strength   agility expertise resistance      life height
2  00100100  00001100  00100110   00010111  00101001    011
3  00100100  00001100  00100110   00010111  00101001    011
4  01001110  00001001  00000011   00010111  00100101    011
5  00100100  00001100  00100110   00010111  00101001    011

In [18]:
from genes import decode_genes
cromosomas_decoded2 = decode_genes(cromosomas_ok)
cromosomas_decoded2

strength  agility  expertise  resistance  life  height
1        36       12         38          23    41     1.6
2        36       12         38          23    41     1.6
3        78        9          3          23    37     1.6
4        36       12         38          23    41     1.6

In [21]:
from character import Character
def eval_performace(generation,type):
    data=pd.DataFrame(generation)
    data['type'] = type
    personajes = []
    for index, row in data.iterrows():
        personaje = Character(row['type'],row['height'],row['strength'],row['agility'],row['expertise'],row['resistance'],row['life'])
        personajes.append(personaje)
    data['performance'] = [personaje.performance for personaje in personajes]

    poblacion_0 = data.sort_values(by='performance', ascending=False)

    return poblacion_0

new_generation = eval_performace(cromosomas_decoded2,type)
new_generation

strength  agility  expertise  resistance  life  height  \
1        36       12         38          23    41     1.6   
2        36       12         38          23    41     1.6   
4        36       12         38          23    41     1.6   
3        78        9          3          23    37     1.6   

                   type  performance  
1  CharacterType.ARCHER   427.263203  
2  CharacterType.ARCHER   427.263203  
4  CharacterType.ARCHER   427.263203  
3  CharacterType.ARCHER    94.169304